cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,443 @@
+        Iterator,                   // the name of the iterator we're adapting
+        thrust::use_default, thrust::use_default, thrust::use_default,
+        T_output,                   // set `super_t::reference` to `T_output`
+        thrust::use_default
+      >
+  {
+  public:
+    // shorthand for the name of the iterator_adaptor we're deriving from
+    using super_t = thrust::iterator_adaptor<
+      column_input_iterator<T_output, T_input, Iterator>, Iterator,
+      thrust::use_default, thrust::use_default, thrust::use_default, T_output, thrust::use_default
+    >;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator(const T_input col, const Iterator &it) : super_t(it), colData(col){}
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator(const T_input col) : super_t(Iterator{0}), colData(col){}
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator(const column_input_iterator &other) : super_t(other.base()), colData(other.colData){}
+
+    // befriend thrust::iterator_core_access to allow it access to the private interface below
+    friend class thrust::iterator_core_access;
+
+  private:
+    const T_input colData;
+
+    // it is private because only thrust::iterator_core_access needs access to it
+    CUDA_HOST_DEVICE_CALLABLE
+    typename super_t::reference dereference() const
+    {
+      gdf_index_type id = *(this->base()); // base() returns base iterator: `Iterator`
+      return colData.at(id);
+    }
+};
+
+// ---------------------------------------------------------------------------
+// helper functions to make iterator
+
+/** -------------------------------------------------------------------------*
+ * @brief helper function to make iterator with nulls
+ * Input iterator which can be used for cub and thrust.
+ *
+ * @tparam T_element The cudf data type of input array
+ * @tparam T_output  The cudf data type of output data or array
+ * @tparam T_output_helper
+ *                   The output helper struct type, usually `ColumnOutputSingle`,
+ *                   `ColumnOutputSquared`, or `ColumnOutputMixed`.
+ * @tparam Iterator_Index
+ *                  The base iterator which gives the index of array.
+ *                  The default is `thrust::counting_iterator`
+ *
+ * @param[in] data     The pointer of column data array
+ * @param[in] valid    The pointer of null bitmask of column
+ * @param[in] identity The identity value used when the mask value is false
+ * @param[in] it       The index iterator, `thrust::counting_iterator` by default
+ * -------------------------------------------------------------------------**/
+template <typename T_element, typename T_output = T_element,
+    typename T_output_helper = ColumnOutputSingle<T_output>,
+    typename Iterator_Index=thrust::counting_iterator<gdf_index_type> >
+auto make_iterator_with_nulls(const T_element *data, const bit_mask::bit_mask_t *valid,
```

Comment:
Instead of having `make_iterator_with_nulls` and `make_iterator_no_nulls`, just add a `bool has_nulls` template argument. 

So instead of `make_iterator_with_nulls`, it's just `make_iterator<true>` or `make_iterator<false>` instead of `make_iterator_without_nulls`.

Reply:
@jrhemstad 
`just add a bool has_nulls template argument` sounds a good idea, but I have a concern.
It is possible to add `bool has_nulls` as 1st template argument, and can create a specialized case make_iterator<true or false>.
However, since all of `make_iterator_xxx` actually  have different parameters.
So the specialization case for each parameter set is only one of <true or false>.
If we don't mind that, I will do it. Do you mind?

Another option (though I don't like this)
Indeed if there is `identity` parameter, it creates null iterator, else, it creates non-null iterator.
We can distinguish null or non-null iterator by `identity` parameter.
But I feel it is ambiguous, I think we should explicitly set null or non-null at creating the iterator.

Reply:
> It is possible to add bool has_nulls as 1st template argument, and can create a specialized case make_iterator.

Since `make_iterator...` has additional template parameters, you're talking about partially specialization, which isn't possible with a free function like this.

I see no need to specialize on the `has_nulls` template parameter. Furthermore, I don't think there is an issue with having unused constructor parameters if `has_nulls==false`. Just default the unused parameters and I think it'll be fine. 

```
template <bool has_nulls, typename T_element, typename T_output = T_element,
    typename T_output_helper = ColumnOutputSingle<T_output>,
    typename Iterator_Index=thrust::counting_iterator<gdf_index_type> >
auto make_iterator(const T_element *data, const bit_mask::bit_mask_t *valid = nullptr,
    T_element identity = T_element{}, Iterator_Index const it = Iterator_Index(0))
```